### PR TITLE
Fully rewritten semantic annotations

### PR DESCRIPTION
**Pull request by mayakwd**
Added lots of checks and fixes, rewritten the old one, added proper models preserving, improved references evaluation.

### CODE REVIEW

**Comment by mayakwd** [comment]
Some leaks in the tests, also I've used new version of method from StringUtils, going to fix it soon. Anyway PR could be reviewed

**Comment by EricBishton** [comment]
You realize that this is going to take a week to review properly, right?  I'll probably review it in stages.

**Comment by EricBishton** [comment]
@mayakwd Hi Ilya, I'm starting to review this now.  Do you want to update it before I go too far?

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 1]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.duplicate.member.name=Duplicate class member declaration ''{0}''
+haxe.semantic.not.allowed.override=Can't override static, inline or final methods
+haxe.semantic.overrides.nothing=Method overrides nothing
+haxe.semantic.duplicate.parameter.name=Duplicate parameter name ''{0}''
+haxe.semantic.invalid.accessor.type=Invalid {0}ter accessor type. Only ''{0}'', ''default'', ''null'', ''never'', ''dynamic'' types allowed
+haxe.semantic.accessor.method.not.found=Accessor method ''{0}'' not found
+haxe.semantic.not.real.var=This field cannot be initialized because it is not a real variable
+haxe.semantic.optional.mark.redundant=Optional mark is redundant when initial value specified
+haxe.semantic.method.visibility.less.then.parent=Method ''{0}'' has less visibility ({1}) than declared at ''{2}'' (was <b>{3}</b>)
+haxe.semantic.not.implemented.members=<html><body>Following members, declared at interface ''{0}'' not implemented:{1}</body></html>
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: ... `The following members, declared at interface ''{0}'' are not implemented: {1}` ...

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 2]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.missing.list=<hr><p><b>{0}</b>: {1}</p>
+haxe.semantic.missing.list.item=''{0}''
+haxe.semantic.constant.value.required=Parameter default type should be constant but was ''{0}''
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: ... `should be constant, but was` ...   (Missing comma.)

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 3]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.interface.cant.implement=No implements clause allowed for interfaces
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: `No 'implements' clause`...

Just because it doesn't read as easily as with the quotes.

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 4]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.not.interface=''{0}'' not an interface
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: `''{0}'' is not an interface.`

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 5]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.should.extend.class=Should extend by using a class
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: Perhaps, `Should extend {0} by using a class.`

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 6]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.should.extend.class.but.extend.0=Should extend by using a class, but extending ''{0}''
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: Perhaps, `Class should extend ''{0}''.`?

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 7]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.field.has.different.accessors=Field has different property access than defined at ''{0}'' (''{1}'' should be ''<b>{2}</b>'')
+haxe.semantic.field.type.required=Field type required by interface ''{0}'', should be ''{1}''
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: Do you mean: `Field type required by interface ''{0}''; should be ''{1}''.` or do you just need to drop the comma?

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 8]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.unexpected.parameter=Unexpected parameter
+haxe.semantic.not.matching.arity=Not matching arity, expected {0} parameters count, got {1}
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: `Arity mismatch:  Expected {0} parameters, but found {1}.`

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 9]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.semantic.remove.parameter=Remove parameter
+haxe.semantic.final.var.init=Final field ''{0}'' must be initialized immediately or in the constructor
+haxe.semantic.parameter.default.constant.required=Parameter default type should be constant but was ''{0}''
 haxe.semantic.final.static.var.init=Static final variable ''{0}'' must be initialized
+haxe.semantic.property.cant.be.final=Property can not be final
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: ...`cannot`...

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 10]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+  </table></body></html>
 
-# HAXE fixes
-haxe.quickfix.make.mutable=Make {0} ''{1}'' mutable
-haxe.quickfix.make.immutable=Make {0} ''{1}'' immutable
-
+# HAXE quickfixes
+# {0} - Found type, {1} - Expected type
+haxe.quickfix.add.cast=Add cast from ''{0}'' to ''{1}''
+haxe.quickfix.remove.excessive.arguments=Remove excessive arguments
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: `Remove excess arguments.`

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 11]
```diff
@@ -171,14 +170,80 @@ haxe.semantic.method.parameter.optional.add=Declare parameter as optional
...
+haxe.quickfix.drop.modifier=Drop modifier ''{0}''
+haxe.quickfix.drop.modifier.from=Drop modifier ''{0}'' from ''{1}''
+haxe.quickfix.drop.extend=Drop extend declaration
+haxe.quickfix.change.type=Change {0} ''{1}'' type from ''{2}'' to ''{3}''
+haxe.quickfix.set.package.name.to=Set package name to ''{0}''
+haxe.quickfix.add.modifier=Add modifier ''{0}''
+haxe.quickfix.setup.module.source.paths=Configure module source-paths
+haxe.quickfix.rename.to=Rename to ''{0}''
+haxe.quickfix.create.accessor.method=Create {1}ter-method ''{0}_{1}''
+haxe.quickfix.add.optional.mark=Remove optional mark
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: `Add optional mark.`?

File: src/common/com/intellij/plugins/haxe/HaxeBundle.properties [hunk 12]
```diff
@@ -187,7 +252,7 @@ haxelib.dependency.list.separator=,
 
 # HAXE Module settings dialog.
 module.settings.synchronize.dependencies=Use Haxe project files  and haxelib to update module dependencies and settings.  (Project file changes update IDEA settings; IDEA does not update Haxe project files.)
-module.settings.synchronize.dependencies.title=Automatically s&ynchronize dependencies and settings
+module.settings.synchronize.dependencies.title=Automatically synchronize dependencies and settings
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/HaxeBundle.properties]: The ampersand is there to set the shortcut for the checkbox to Alt+Y.

File: src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java [hunk 13]
```diff
@@ -188,624 +253,706 @@ private static boolean isFieldInitializedInTheConstructor(HaxeFieldModel field)
...
-}
 
-class MethodBodyChecker {
-  public static void check(HaxeMethod psi, AnnotationHolder holder) {
-    final HaxeMethodModel method = psi.getModel();
-    HaxeTypeResolver.getPsiElementType(method.getBodyPsi(), holder);
+  private static boolean isSingleQuotesRequired(HaxeStringLiteralExpression o) {
+    if (o.getParent() instanceof HaxeObjectLiteralElement) return false;
+    return (o.getLongTemplateEntryList().size() > 0 || o.getShortTemplateEntryList().size() > 0) &&
+           o.getFirstChild().textContains('"');
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: Use `o.getFirstChild().getText().startsWith('"');`. Otherwise, escaped quotes will be detected.

File: src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java [hunk 14]
```diff
@@ -188,624 +253,706 @@ private static boolean isFieldInitializedInTheConstructor(HaxeFieldModel field)
...
   }
 
-  @Override
-  public void visitElement(PsiElement element) {
-    super.visitElement(element);
-    if (result) return;
-    if (element instanceof HaxeIdentifier || element instanceof HaxePsiToken || element instanceof HaxeStringLiteralExpression) return;
-    element.acceptChildren(this);
+  private String wrapWithHtmlError(String string) {
+    String color = UIUtil.isUnderDarcula() ? "FF6B68" : "red";
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java]: This color should be in the settings dialog.  It's OK  to create a "technical debt" issue for this one, because it's going to take UI work that I'd rather see done in a separate PR.

File: src/common/com/intellij/plugins/haxe/ide/quickfix/HaxeCreateAccessorMethodFix.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+      PsiElement anchor = classBody.getLastChild();
+
+      for (PsiElement element : elements) {
+        PsiElement newLine = HaxeElementGenerator.createNewLine(project);
+        anchor = classBody.addAfter(element, anchor);
+        anchor = classBody.addBefore(newLine, anchor);
+      }
+    }
+  }
+
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/ide/quickfix/HaxeCreateAccessorMethodFix.java]: Do you want an isAvailable() that checks whether a class body can be found?

File: src/common/com/intellij/plugins/haxe/ide/quickfix/HaxeCreateAccessorMethodFix.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    HaxePsiCompositeElement classBody = declaringClass != null ? declaringClass.getBodyPsi() : null;
+    if (classBody != null) {
+      final StringBuilder builder = new StringBuilder();
+      GetterSetterMethodBuilder.build(builder, model, accessorType == HaxeAccessorType.GET);
+
+      List<HaxeNamedComponent> elements =
+        HaxeElementGenerator.createNamedSubComponentsFromText(project, builder.toString());
+
+      PsiElement anchor = classBody.getLastChild();
```
Comment (EricBishton) [inline src/common/com/intellij/plugins/haxe/ide/quickfix/HaxeCreateAccessorMethodFix.java]: Throw an `// TODO: Use (or define) style setting to determine order.` comment in here, so it will be easier to find.  For now, putting them in the last place is OK.
